File: app/models/forum.rb
#Copyright 2012 Vishal Parikh
#This file is part of Freader.
#Freader is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.
#
#Freader is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with Freader.  If not, see <http://www.gnu.org/licenses/>.

# == Schema Information
#
# Table name: forums
#
#  id         :integer         not null, primary key
#  title      :string(255)
#  sid        :string(255)
#  created_at :datetime
#  updated_at :datetime
#

class Forum < ActiveRecord::Base
  include SessionsHelper
  attr_accessible :title

  ################################### Associations ###############################
  has_many :posts
  has_many :comments, :through=>:posts
  has_many :subscriptions
  has_many :users, :through=>:subscriptions
  has_one :owner, :class_name => 'Subscription', :conditions => {:status => 'owner'}

  ################################### Validations ################################
  before_validation {self.sid ||= gen_token}

  validates :title, :presence=>true
  validates :sid, :presence=>true, :uniqueness=>true

  def permission(user, thing)
    #right now, anyone can view
    if(thing==:view)
      return true
    end

    if(user.nil?)
      return false
    end
    sub = Subscription.find_by_user_id_and_forum_id(user.id, self.id)
    if(sub.nil?)
      return false
    end
    #anyone with a subscription can post
    return true
  end

  def make_markers_for(post)
    if(post.forum.id != id)
      throw "invalid post"
    end

    self.users.each do |u|
      m = Marker.new(:user_id=>u.id, :post_id=>post.id, :forum_id=>self.id, :is_read=>false, :is_starred=>false, :is_hidden=>false)
      if(u.id==post.user_id) #this is the owner
        m.is_read=true
      end
      #TODO more graceful error handling
      m.save!
    end
  end

  def mark_all_read(user)
    Marker.where(:forum_id=>id, :user_id=>user.id).find_each{|m|m.update_attributes(:is_read=>true)}
  end

  def as_json(options)
    options||={}
    #TODO include subscribers
    #TODO prefetch
    ret = {:title=>title, :id=>sid, :owner=>owner ? owner.user.sid : nil}
    if(options[:current_user])
      sub = Subscription.find_by_user_id_and_forum_id(options[:current_user].id, id)
      ret[:subscription] =sub 
      ret[:unread_count] = Marker.where(:forum_id=>id, :user_id=>options[:current_user].id, :is_read=>false).count
    end
    if(options[:include_posts])
      offset = options[:offset] || 0
      limit = options[:limit] || 20
      if(options[:current_user])
        #get unread posts
        if(options[:hide_read])
          m = Marker.where(:forum_id=>self.id, :user_id=>options[:current_user].id, :is_read=>false).order("created_at DESC").offset(offset).limit(limit).all
        else
          m = Marker.where(:forum_id=>self.id, :user_id=>options[:current_user].id).order("created_at DESC").offset(offset).limit(limit).all
#        if(m.length<limit && !options[:hide_read])
#          m+=Marker.where(:forum_id=>self.id, :user_id=>options[:current_user].id, :is_read=>true).offset([offset-m.length, 0].max).limit(limit-m.length).all
        end
        ret[:posts] = Post.where("id IN (?)", m.map{|k|k.post_id}).includes(:comments).as_json(options)
      else
        ret[:posts] = self.posts.order("updated_at DESC").offset(offset).limit(limit).includes(:comments).as_json(options)
      end
    end
    return ret
    
  end


end

File: public/javascripts/models/forum.js
/*
Copyright 2012 Vishal Parikh                                             
This file is part of Freader.                                            
Freader is free software: you can redistribute it and/or modify          
it under the terms of the GNU General Public License as published by     
the Free Software Foundation, either version 3 of the License, or        
(at your option) any later version.                                      
                                                                         
Freader is distributed in the hope that it will be useful,               
but WITHOUT ANY WARRANTY; without even the implied warranty of           
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            
GNU General Public License for more details.                             
                                                                         
You should have received a copy of the GNU General Public License        
along with Freader.  If not, see <http://www.gnu.org/licenses/>.         
*/
/*global App Backbone _ GlobalSettings*/
App.Models.Forum = Backbone.Model.extend({
    //introspection
    _class:'Forum',
    _members:['posts'],

    urlRoot: 'forums',
    //override url function to include 'prefetch' option
    url: function() {
      var base = this.urlRoot;
      if (this.isNew()) return base;
      var u =  base + (base.charAt(base.length - 1) === '/' ? '' : '/') + encodeURIComponent(this.id);
      if(this.prefetch){
        return u + '?prefetch=true';
      }else{
        return u;
      }
    },
    prefetch:'false', //if this is true, when we fetch, we also fetch the posts
    initialize:function(){
      _.bindAll(this, 'url');
    },

    markAllRead:function(){
      this.save(null, {url:this.url().replace(/\?|$/,'/mark_all_read?')});
      this.posts().forEach(function(p){p.set({"is_read":true});});
    },

    getMore:function(count, f){
      count=count||20;
      this.fetch({data:{offset:this.posts().length, limit:count}, success:f});
    },

    posts:function(){
      if(!this._posts){this._posts=new App.Collections.Posts(null,{forum:this});}
      return this._posts;
    },

    //override parse so we can handle a response that includes posts
    //the resp objection should structured as follows 
    //forum=><forum attributes>
    //posts=>[{post=><post attributes>, comments=>[{comment=>{comment attributes}}, ...]},
    //        {post=><post attributes>, comments=>[{comment=>{comment attributes}}, ...]},
    //        ...
    //       ]
    parse:function(resp){
      var f, p, v;
      if(resp.forum){
        f = resp.forum;
        p = resp.forum.posts;
        v = resp.version;
      }else{
        f=resp;
      }
      if(v && v!==GlobalSettings.version){
        App.notifier.notify("A new version is available, please reload the page", {type:'success', expire:0});
      }
      if(p){
        console.log("got "+p.length+"posts");
        var merged = this.posts().merge(p, {parse:true});
      }

      return f;
    }

});

App.Collections.Forums = Backbone.Collection.extend({
  model:App.Models.Forum,
});
